src/Compilers/CSharp/Test/Semantic/Utilities/ValueSetTests.cs

```diff
@@ -0,0 +1,663 @@
+                test(Random.Next(int.MinValue, int.MaxValue));
+                test(Random.Next(int.MinValue, int.MaxValue));
+                void test(int val)
+                {
+                    Assert.Equal(val >= i1 && val <= i2, values.Any(Equal, val));
+                    Assert.Equal(val >= i1, values.Any(LessThanOrEqual, val));
+                    Assert.Equal(val > i1, values.Any(LessThan, val));
+                    Assert.Equal(val <= i2, values.Any(GreaterThanOrEqual, val));
+                    Assert.Equal(i2 > val, values.Any(GreaterThan, val));
+                }
+            }
+        }
+
+        [Fact]
+        public void TestIsEmpty_01()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                int i1 = Random.Next(int.MinValue, int.MaxValue);
+                int i2 = Random.Next(int.MinValue, int.MaxValue);
+                IValueSet<int> values = ForInt.Related(GreaterThanOrEqual, i1).Intersect(ForInt.Related(LessThanOrEqual, i2));
+                Assert.Equal(values.ToString().Length == 0, values.IsEmpty);
+            }
+        }
+
+        [Fact]
+        public void TestDouble_01()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                double d1 = Random.NextDouble() * 100 - 50;
+                double d2 = Random.NextDouble() * 100 - 50;
+                if (d1 > d2) (d1, d2) = (d2, d1);
+                IValueSet<double> values = ForDouble.Related(GreaterThanOrEqual, d1).Intersect(ForDouble.Related(LessThanOrEqual, d2));
+                Assert.Equal(FormattableString.Invariant($"[{d1:G17}..{d2:G17}]"), values.ToString());
+            }
+        }
+
+        [Fact]
+        public void TestChar_01()
+        {
+            IValueSet<char> gea1 = ForChar.Related(GreaterThanOrEqual, 'a');
+            IValueSet<char> lez1 = ForChar.Related(LessThanOrEqual, 'z');
+            IValueSet<char> gea2 = ForChar.Related(GreaterThanOrEqual, 'A');
+            IValueSet<char> lez2 = ForChar.Related(LessThanOrEqual, 'Z');
+            var letters = gea1.Intersect(lez1).Union(gea2.Intersect(lez2));
+            Assert.Equal("['A'..'Z'],['a'..'z']", letters.ToString());
+        }
+
+        [Fact]
+        public void TestDouble_02()
+        {
+            Assert.Equal("-Inf", ForDouble.Related(LessThan, double.MinValue).ToString());
+            var lt = ForDouble.Related(LessThan, 0.0);
+            Assert.Equal(FormattableString.Invariant($"-Inf,[{double.MinValue:G17}..{-double.Epsilon:G17}]"), lt.ToString());
+            var gt = ForDouble.Related(GreaterThan, 0.0);
+            Assert.Equal(FormattableString.Invariant($"Inf,[{double.Epsilon:G17}..{double.MaxValue:G17}]"), gt.ToString());
+            var eq = ForDouble.Related(Equal, 0.0);
+            Assert.Equal("[0..0]", eq.ToString());
+            var none = lt.Complement().Intersect(gt.Complement()).Intersect(eq.Complement());
+            Assert.Equal("NaN", none.ToString());
```

Comment:
>NaN [](start = 26, length = 3)

Why is the result "NaN" rather than an empty set?

Reply:
Because NaN is a member of all three sets.

---
In reply to: [383432386](https://github.com/dotnet/roslyn/pull/41503#discussion_r383432386) [](ancestors = 383432386)